Feat node revamp deployment (#27984)
TheOdinProject/curriculum: nodeJS/express/deployment.md

- Know how to troubleshoot common deployment issues.

### What are hosting providers?

Hosting providers are like server landlords. They own servers and rent space on them to customers, who can then use the space to store their websites and make them accessible to anyone on the web.

You've already had some experience using a hosting provider when you deployed projects to Github pages earlier in the curriculum. GitHub pages is great for hosting **static** web pages for free, but we won't be able to use it for hosting our **dynamic** NodeJS apps. We're going to need something more powerful.

### Static vs dynamic sites

Static websites consist of pre-written HTML pages. They are "static" because everyone who visits them will see the same content. To build static sites, you only need HTML, CSS and JavaScript.

Dynamic websites, on the other hand, are websites that can change content based on the user who is visiting them. X (formerly known as Twitter) is a good example; every user sees different content on their homepage feed based on who they follow. To build dynamic sites, you still need HTML, CSS and JS. But additionally, you need a server-side application and a database.

> This additional tech prohibits us from using GitHub Pages for hosting our NodeJS apps. GitHub pages cannot run NodeJS applications and doesn't have database services we can use.

Reviewer: One last thing from me, just adding this in to extend the PaaS justification to cover "why not use Netlify/Vercel?" which is a somewhat common question for new starters.

```suggestion
This additional tech prohibits us from using GitHub Pages for hosting our NodeJS apps. GitHub pages cannot run Node.js applications and doesn't have database services we can use. Similarly, host providers like Netlify and Vercel, which you may have used back in the React course, do not have the same capabilities to run our Node.js servers and databases. They're not the right tools for our back ends.
```

We will provide a list of our recommended PaaS providers later in the lesson. First, let us explore from a high level how PaaS providers work.

### How do PaaS services work?

PaaS providers work by giving you easy access to a few resources that any NodeJS app can't live without to function on the web.

#### Instances

The first and most crucial thing PaaS providers give you are virtual "computers" called instances which run your app. Basically, one instance means a single instance of your application running at one time. That's like having a single computer run your app like you do on Localhost. Multiple instances are like having several copies of your app running simultaneously, which allows you to handle more traffic.

For most of your apps, one instance is plenty enough. You can support a lot of traffic using just a single instance. Many of the PaaS providers we will recommend later in this lesson give you your first one for free.

<div class="lesson-note" markdown="1">

> Learners can keep their server instance and database instance on one PaaS or use separate PaaS's to reduce hosting costs!

Reviewer: Reframing this note since the original sounded like we want learners to deal with paid stuff during this course.

```suggestion
Learners can keep their server instance and database instance on one PaaS or use separate PaaSes if necessary. When you deal with paid plans, it might even reduce hosting costs!
```

#### Going further with troubleshooting tools

As your application grows, you'll want to get more sophisticated with your error-tracking tools. For example, you can use services like [Sentry](https://sentry.io/) to track and monitor errors using a slick, easy-to-use interface and get notified when they happen.

These services will give you more information about the error and the request that caused it, saving you a ton of time. However, setting up and using these services are out of the scope of this lesson. You can get by just fine with the logs for your first few apps.

#### One final tip

If something has broken in your latest deployment after successful deployments in the past, backtrack to the last working version to determine what changes you made and slowly reintroduce those changes again if you need to.

This will be where the Git skills you've been learning will start to really pay off and save you an immense amount of time. You'll be able to use `git log` to see the history of your latest changes and `git checkout` to revert to a previous working version quickly.

### Assignment

<div class="lesson-content__panel" markdown="1">

> 1. Deploy your [Mini Message Board project](https://www.theodinproject.com/lessons/nodejs-mini-message-board) to one of the hosting providers we've mentioned. If you need help deciding which one to use, we recommend Fly.io. The important thing to take away from this first deployment is getting experience deploying. Don't worry if you don't understand everything that's happening. That will come with time.

Reviewer: Using the [current deployment lesson](https://www.theodinproject.com/lessons/nodejs-deployment#assignment)'s text, as I think when you branched, you would've copied over the text from a now outdated version of the lesson.

```suggestion
1. Deploy your [Mini Message Board project](https://www.theodinproject.com/lessons/nodejs-mini-message-board) to one of the hosting providers we've mentioned. Any of the free options will work for curriculum purposes, so it doesn’t matter which you pick for your projects. The important thing to take away from this first deployment is getting experience deploying. Don’t worry if you don’t understand everything that’s happening. That will come with time.
```

   - Use one of the linked deploy guides for your PaaS provider to help you through the process.
   - If you're having trouble deploying, check out the [Debugging and Troubleshooting Deployments](#debugging-and-troubleshooting-deployments) section for some tips.

>    - If for some reason the deployment is still too difficult, and you just can't seem to get it to work, move onto the next project and come back once you've deployed your MDN project. The steps in their tutorial hold your hand a bit more through the process and will give you the confidence you need to deploy this project.

Reviewer: We need to get rid of this last bullet since we no longer have the MDN tutorial to refer to. No need for a replacement.

```suggestion
```

Luckily, there are still plenty of other great options out there. The downside is that they all have very limited free tiers. For this reason, and to accommodate as many of our learners as possible, we're going to recommend a range of options instead of just one.

You can host most of your projects for free using a combination of the different providers we recommend, but you'll have a bit more leg work to do. You'll have to create an account and get familiar with a few different providers.

If you can pay for hosting, things are more straightforward. It will afford you the luxury of choosing one provider you can learn deeply, and you'll be able to manage all your apps from one place.

Whatever your circumstances, we've got you covered. Here are the PaaS providers we recommend.

---

#### Fly.io

- Fly.io uses a convenient CLI tool for deployment.
- Pay for what you use with very reasonable rates. Each app should cost around $4 per month.
- $20 a month should be enough to host eight apps (including three apps for free).
- Fly.io charges $0.15/GB of RootFS for machines stopped for 30 days.

> ##### Fly.io: Free Plan
> 
> - You can host three apps for free before you need to start paying.
> - Fly.io waives monthly invoices that total less than $5 USD. So, although it may look like
>   you are being charged, it's very unlikely that you will exceed $5 and actually have to pay.

Reviewer: Could the verbiage for this part be changed to match the updated text in #28036 (inc. pending changes from comments)? Just the hobby plan bit, not the legacy hobby plan bit.

```suggestion
##### Fly.io: Hobby Plan

- New customers get a one-time $5 free trial credit to test Fly.io at no cost. The credit will only start being used if you exceed the free resource allowance, and once it's been used up, you will be automatically placed on the $5/month Hobby plan.
- The longevity of your free trial credit depends on how many resources you consume. More complex apps with more traffic may consume all free resources within a month, whereas simpler apps may last longer.
```

Reviewer: **Update:** #28199 made further changes to the Fly.io section, due to *even more* recent changes to their free trial. The current lesson content for Fly.io is the latest, so if you could use the text from the current lesson, that'd be great!

##### Render: Free Plan

- 750 hours of free usage per month.
- Applications are put to sleep automatically after 15 minutes of inactivity, so the 750 free hours should be enough to host a few apps for the entire month.
- Requires a credit card.

##### Render: Links

- [Render homepage](https://render.com/)
- [Render documentation](https://render.com/docs/)
- [Guide: Official getting started with Node/Express on Render guide](https://render.com/docs/deploy-node-express-app)
- [Guide: CodeBrah video guide for deploying NodeJS applications to Render](https://www.youtube.com/watch?v=bnCOyGaSe84&ab_channel=CodeBrah)

---

#### Koyeb

- Easily Git push to deploy.
- Ability to run web apps, APIs, and workers easily.
- SSL and 10 free custom domains

##### Koyeb: Free Plan

- One free Web service.
- One free Postgres database (50 hours).
- Doesn't require a credit card to start.
- Pay-per-use by the second for more.

##### Koyeb: Links

- [Koyeb homepage](https://www.koyeb.com/)
- [Koyeb documentation](https://www.koyeb.com/docs)
- [Guide: Official getting started with Node/Express on Koyeb guide](https://www.koyeb.com/docs/deploy/express)

---

#### Neon

- Database branching

> -

Reviewer: Rogue line?
```suggestion
```

Author: Yea basically a placeholder to indicate more information should probably go here
